### ISSUE TITLE
【Task.17】Replace the DeepEyes processor file override with dynamic registration

### ISSUE THREAD

**Issue by A-Words** [redai-infra/Relax#163]
## Background and Goal

Official task: [17 DeepEyes processor dynamic patch](https://github.com/redai-infra/community/blob/main/contributor-program/2026-cohort-1/official-task.md#17-deepeyes-processor-%E5%8A%A8%E6%80%81-patch)

Claim: [A-Words' registration comment in redai-infra/Relax#86](https://github.com/redai-infra/Relax/issues/86#issuecomment-5078060349)

`examples/deepeyes/run_deepeyes_r3.sh` currently copies the repository's full `qwen_vl.py` into SGLang's `/sgl-workspace` installation directory. This depends on a fixed image path and requires Relax to maintain a processor copy that can drift from upstream.

This task will use SGLang's external package registration mechanism instead. It will inject only the behavior DeepEyes needs for pre-expanded image tokens while continuing to reuse the upstream implementation for the remaining image-processing logic.

## Fixed Information

| Item | Value |
|---|---|
| Upstream repository | `redai-infra/Relax` |
| Development fork | `A-Words/Relax` |
| Base branch | `main` |
| Base commit | `b9da9f818fca70ff311009d4213be196acfa426e` |
| Development branch | `fix/deepeyes-dynamic-processor-patch` |
| Delivery | This Issue + a Draft PR |
| Container image | `ghcr.io/redai-infra/relaxrl:dev-20260723-8cc1e8fd` |
| Python / PyTorch / CUDA | 3.12.3 / 2.11.0+cu129 / 12.9 |
| SGLang / Ray / Transformers | 0.5.12.post1 / 2.56.1 / 5.6.0 |
| Smoke GPU | 1 x NVIDIA A100-SXM4-80GB |
| CPU / memory / disk | AMD EPYC 7763; 1.0 TiB RAM; 256 GB container disk |

## Entry Point and Full Command

The existing repository script remains the official entry point. `NUM_ROLLOUT=1` is used only for the acceptance smoke; the script retains the model, data, and other training parameters:

```bash
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export NUM_GPUS=8
export MODEL_DIR=/path/to/models
export DATA_DIR=/path/to/data
export SAVE_DIR=/path/to/save
export NUM_ROLLOUT=1

bash examples/deepeyes/run_deepeyes_r3.sh
```

The only parameter change to the official entry point is:

```text
--sglang-external-model-package examples.deepeyes.sglang_patch
```

### Acceptance Smoke Command

The following command freezes the reduced single-GPU acceptance configuration. It changes the model size, resource topology, data slice, and batch/rollout sizes required for a one-GPU smoke; it is not an official training recipe. Actual results and evidence belong in the Draft PR.

```bash
export CUDA_VISIBLE_DEVICES=0
export NUM_GPUS=1
export MODEL_DIR=/root/exp/models
export DATA_DIR=/root/exp/data
export SAVE_DIR=/root/exp/save
export PROJECT_NAME=Relax/task17-smoke
export DEEPEYES_JUDGE_API_KEY=EMPTY
export DEEPEYES_JUDGE_BASE_URL=http://127.0.0.1:30000/v1
export DEEPEYES_JUDGE_MODELS=Qwen2.5-0.5B-Instruct

source scripts/entrypoint/local.sh

# Propagate the judge configuration to Ray workers.
export RUNTIME_ENV_JSON="$(python3 - <<'PY'
import json
import os

runtime_env = json.loads(os.environ["RUNTIME_ENV_JSON"])
runtime_env["env_vars"].update(
    {
        "DEEPEYES_JUDGE_API_KEY": os.environ["DEEPEYES_JUDGE_API_KEY"],
        "DEEPEYES_JUDGE_BASE_URL": os.environ["DEEPEYES_JUDGE_BASE_URL"],
        "DEEPEYES_JUDGE_MODELS": os.environ["DEEPEYES_JUDGE_MODELS"],
    }
)
print(json.dumps(runtime_env))
PY
)"

python -m sglang.launch_server \
  --model-path /root/exp/models/Qwen2.5-0.5B-Instruct \
  --port 30000 \
  --api-key EMPTY \
  --mem-fraction-static 0.05 \
  > logs/task17-judge.log 2>&1 &
JUDGE_PID=$!
trap 'kill "$JUDGE_PID" 2>/dev/null || true' EXIT

# Wait for http://127.0.0.1:30000/health to return HTTP 200 before submitting
# the Ray job below.
```

<details>
<summary>Complete effective Ray argv</summary>

```bash
ray job submit --runtime-env-json="${RUNTIME_ENV_JSON}" \
  --address=http://127.0.0.1:8265 -- \
python3 relax/entrypoints/train.py \
  --rollout-num-gpus-per-engine 1 \
  --resource '{"actor": [1, 1], "rollout": [1, 1]}' \
  --max-staleness 0 --num-data-storage-units 1 --colocate \
  --swiglu --num-layers 36 --hidden-size 2560 --ffn-hidden-size 9728 \
  --num-attention-heads 32 --group-query-attention --num-query-groups 8 \
  --use-rotary-position-embeddings --disable-bias-linear \
  --normalization RMSNorm --norm-epsilon 1e-6 --rotary-base 5000000 \
  --vocab-size 151936 --kv-channels 128 --qk-layernorm \
  --hf-checkpoint /root/exp/models/Qwen3-VL-4B-Instruct \
  --ref-load /root/exp/models/Qwen3-VL-4B-Instruct \
  --save /root/exp/save/Qwen3-VL-4B-Instruct-Checkpoint \
  --megatron-to-hf-mode bridge --save-interval 100 --max-actor-ckpt-to-keep 3 \
  --prompt-data "['/root/exp/data/deepeyes-v1/data_0.1.2_visual_toolbox_v2.parquet@[0:1]']" \
  --input-key prompt --label-key reward_model \
  --multimodal-keys '{"image":"images"}' --reward-key score --metadata-key extra_info \
  --apply-chat-template \
  --custom-generate-function-path examples.deepeyes.rollout.generate \
  --custom-rm-path examples.deepeyes.reward_deepeyes.reward_func \
  --custom-config-path examples/deepeyes/deepeyes_config.yaml \
  --num-rollout 1 --rollout-batch-size 1 --micro-batch-size 1 \
  --n-samples-per-prompt 1 --rollout-max-response-len 512 \
  --rollout-max-prompt-len 1024 --rollout-temperature 1 --global-batch-size 1 \
  --rollout-shuffle --advantage-estimator grpo \
  --kl-loss-coef 0.00 --kl-loss-type low_var_kl --entropy-coef 0.00 \
  --eps-clip 0.2 --eps-clip-high 0.28 --eps-clip-c 3 --use-tis \
  --optimizer adam --lr 1e-6 --lr-decay-style constant --weight-decay 0.1 \
  --adam-beta1 0.9 --adam-beta2 0.98 --optimizer-cpu-offload \
  --overlap-cpu-optimizer-d2h-h2d --use-precision-aware-optimizer --no-rope-fusion \
  --sglang-mem-fraction-static 0.60 \
  --sglang-external-model-package examples.deepeyes.sglang_patch \
  --sglang-disable-cuda-graph \
  --use-tensorboard --use-metrics-service \
  --tb-project-name Relax/task17-smoke --tb-experiment-name task17-smoke \
  --tensor-model-parallel-size 1 --sequence-parallel \
  --pipeline-model-parallel-size 1 --context-parallel-size 1 \
  --expert-model-parallel-size 1 --expert-tensor-parallel-size 1 \
  --recompute-granularity full --recompute-method uniform --recompute-num-layers 1 \
  --max-tokens-per-gpu 2048 --attention-dropout 0.0 --hidden-dropout 0.0 \
  --accumulate-allreduce-grads-in-fp32 --attention-softmax-in-fp32 \
  --attention-backend flash --use-dynamic-batch-size --skip-eval-before-train \
  --eval-interval 100 \
  --eval-prompt-data vstar /root/exp/data/deepeyes-v1/data_0.1.2_visual_toolbox_v2.parquet@[0:1] \
  --n-samples-per-eval-prompt 8 --eval-max-response-len 2048 --eval-top-p 0.7
```

</details>

## Proposed Implementation

1. Delete `examples/deepeyes/qwen_vl.py` and remove the `cp` command and `/sgl-workspace` hardcoding from the script.
2. Add the `examples.deepeyes.sglang_patch` external package and load it through the existing `--sglang-external-model-package` option so SGLang registers the processor at startup.
3. Make `DeepEyesQwenVLImageProcessor` inherit the upstream `QwenVLImageProcessor` and override only `process_mm_data_async()`:
   - collapse consecutive `<|image_pad|>` tokens in pre-expanded `list[int]` input;
   - call the upstream processor once;
   - restore the caller's original `input_ids`;
   - when multi-turn tool use produces multiple image items, merge their grids before recomputing mRoPE;
   - preserve all other upstream dataclass fields.
4. At startup, require SGLang `0.5.12.post1` and validate the relevant method signatures. Fail fast with upgrade guidance when the contract is incompatible.
5. Add CPU unit tests and SGLang upgrade guidance to the README.

## Scope

In scope:

- `examples/deepeyes/run_deepeyes_r3.sh`
- `examples/deepeyes/processor_patch_utils.py`
- `examples/deepeyes/sglang_patch/`
- `examples/deepeyes/README.md`
- `tests/examples/test_deepeyes_processor_patch.py`
- deletion of `examples/deepeyes/qwen_vl.py`

Out of scope:

- changing the shared Relax SGLang backend or its argument definitions;
- changing the DeepEyes rollout, reward, tool, or training algorithm;
- copying upstream image preprocessing, mRoPE formulas, or other processor implementations;
- restoring any logic that writes into the SGLang installation directory;
- committing the small-model smoke configuration as an official training recipe.

### Baseline and Targets

| Criterion | Baseline `b9da9f8` | Target |
|---|---|---|
| Installation override | A 454-line `qwen_vl.py` copy; one `cp`; one `/sgl-workspace` reference | Remove the copy, override command, and installation-path hardcoding |
| Patch scope | Maintain and overwrite the full upstream processor file | Override only the necessary method in a subclass and reuse upstream for everything else |
| Repeated imports/calls | No dedicated guard or test | Repeated import, registration, and invocation do not accumulate state, mutate caller input, or change outputs |
| Incompatible SGLang | No startup guard | Fail fast during startup when the version or critical method signatures do not match |
| Critical output consistency | The legacy patch preserves pre-expanded `input_ids` and recomputes mRoPE | Preserve the upstream output type, `mm_items`, and unaffected fields while keeping pre-expanded `input_ids` and mRoPE semantics consistent |
| Processor patch tests | 0 | Cover pre-expanded tokens, single/multi-image mRoPE, idempotency, bypass paths, and fail-fast behavior |
| Upgrade guidance | No processor compatibility or upgrade guidance in the README | Document the supported SGLang version, upgrade checks, and incompatible-contract behavior |
| DeepEyes E2E smoke | No task-specific baseline | Complete at least one rollout/train step and evaluation with exit code 0 and no unexplained errors or resource leaks |

This is a correctness and maintainability task, so it does not set a before/after throughput target. Resource and throughput measurements from the smoke are diagnostic only and are not claimed as performance improvements.

- [ ] The DeepEyes directory no longer contains an installation override command or `/sgl-workspace` hardcoding
- [ ] The dynamic processor overrides only the necessary method
- [ ] Repeated imports/calls are stateless, and text input remains on the upstream path
- [ ] Unsupported SGLang versions or method signatures fail fast
- [ ] The upstream output type, `mm_items`, and unaffected fields are preserved
- [ ] Targeted tests cover pre-expanded token/mRoPE behavior for single-image and multi-turn multi-image inputs
- [ ] The README documents the compatible version and upgrade procedure
- [ ] The DeepEyes E2E smoke completes rollout, tool use, reward, a training step, evaluation, and cleanup

For E2E acceptance, the relevant measures are process exit status, HTTP request results, rollout/train/eval completion markers, exception scanning, and post-exit GPU-memory cleanup. Model reward is recorded as trajectory output but is not used as a quality-improvement metric for this patch.
